flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java

```diff
@@ -300,6 +306,240 @@ public void testKafkaTableWithMultipleTopics() throws Exception {
+                                + "  `physical_2` INT,\n"
+                                // metadata fields are out of order on purpose
+                                // offset is ignored because it might not be deterministic
+                                + "  `timestamp-type` STRING METADATA VIRTUAL,\n"
+                                + "  `timestamp` TIMESTAMP(3) METADATA,\n"
+                                + "  `leader-epoch` INT METADATA VIRTUAL,\n"
+                                + "  `headers` MAP<STRING, BYTES> METADATA,\n"
+                                + "  `partition` INT METADATA VIRTUAL,\n"
+                                + "  `topic` STRING METADATA,\n"
+                                + "  `physical_3` BOOLEAN\n"
+                                + ") WITH (\n"
+                                + "  'connector' = 'kafka',\n"
+                                + "  'topic' = '%s',\n"
+                                + "  'properties.bootstrap.servers' = '%s',\n"
+                                + "  'properties.group.id' = '%s',\n"
+                                + "  'scan.startup.mode' = 'earliest-offset',\n"
+                                + "  %s\n"
+                                + ")",
+                        topic, bootstraps, groupId, formatOptions());
+
+        tEnv.executeSql(createTable);
+
+        String initialValues =
+                String.format(
+                        "INSERT INTO kafka\n"
+                                + "VALUES\n"
+                                + " ('data 1', 1, TIMESTAMP '2020-03-08 13:12:11.123', MAP['k1', X'C0FFEE', 'k2', X'BABE'], '%s', TRUE),\n"
+                                + " ('data 2', 2, TIMESTAMP '2020-03-09 13:12:11.123', CAST(NULL AS MAP<STRING, BYTES>), '%s', FALSE),\n"
+                                + " ('data 3', 3, TIMESTAMP '2020-03-10 13:12:11.123', MAP['k1', X'10', 'k2', X'20'], '%s', TRUE)",
+                        topic, topic, topic);
+        try {
+            tEnv.executeSql(initialValues).await();
+            fail(
+                    "Unable to create the Kafka sink table with table option 'topic' and metadata column 'topic'.");
+        } catch (Exception e) {
+            assertTrue(e instanceof ValidationException);
+            assertEquals(
+                    String.format(
+                            "Invalid metadata key '%s' in column 'topic' of table 'default_catalog.default_database.kafka'. "
+                                    + "The %s class '%s' supports the following metadata keys for writing:\n%s",
+                            TOPIC.key,
+                            DynamicTableSink.class.getSimpleName(),
+                            KafkaDynamicSink.class.getName(),
+                            String.join("\n", Arrays.asList(HEADERS.key, TIMESTAMP.key))),
+                    e.getMessage());
+        }
+
+        // ------------- cleanup -------------------
+
+        deleteTestTopic(topic);
```

Comment:
We should make sure this will always be called as soon as `createTestTopic` has been called. Right now there's a few things that could throw exceptions in between and would prevent this from being called.

Reply:
@Airblader , the unit tests added could ensure the `deleteTestTopic(topic)` called after `createTestTopic`.

Reply:
This was marked as resolved, but I don't see any changes here. If there's a failure or a wrong assertion, the cleanup won't be called.

Reply:
@Airblader , I replied to the comment above and no any things could prevent `deleteTestTopic` being called in the added test cases.

Reply:
Almost everything here can prevent `deleteTestTopic` from being called. `tEnv.executeSql` can throw an exception, any assertion which fails would prevent it from being called (since failing assertions throw exceptions). Basically this all needs to be wrapped into try-finally:

```
try {
  createTestTopic();

 // …
} finally {
  deleteTestTopic();
}
```

But I'm also just noticing that all the existing tests in this class have the same issue. :-(

Reply:
@Airblader , does this problem need to be fixed in this pull request?

Reply:
I think it'd be good if we can at least fix it for the new tests you added, but we don't need to change the existing tests as well.

Reply:
@Airblader , I would like to wrap the new tests added in the pull request.